spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkCatalog.java

```diff
@@ -370,13 +371,21 @@ public void alterNamespace(String[] namespace, NamespaceChange... changes) throw
   }
 
   @Override
-  public boolean dropNamespace(String[] namespace) throws NoSuchNamespaceException {
+  // Spark assumes that catalogs CASCADE by default. So we have to eagerly
+  // attempt to drop namespaces and tables, but the CASCADE keyword is still
+  // required to actually drop tables and namespaces.
+  //
+  // TODO - Somehow this got rid of the checked NoSuchNamespaceException.
+  //        I don't think we should check for it and throw if it's not present,
+  //        as we don't know if the user used `if not exists` or not.
```

Comment:
Should look into why this isn't thrown anymore.... Given that we're relying on Spark to handle the `ifExists` type behavior and we don't know if the user used that, it doesn't quite make sense to just check if the `namespace` exists or not.

But it does seem like this should be a possibility. I think it's just no longer throwing _our_ `NoSuchNamespaceException`, but the one from Spark (which comes from Scala and thus isn't checked).

Reply:
What if user drop same namespace cascade concurrently? Maybe we should keep `NoSuchNamespaceException`

Reply:
The reason that I removed it is that IntelliJ told me that it wasn't a used checked exception (i.e. IntelliJ told me that it wouldn't be thrown).

The `NoSuchNamespaceException` in this case is something that I believe bubbles up from Spark.

I'm happy to add the annotation back, but IntellliJ reported that it wasn't thrown.

Reply:
It's different for the SparkSessionCatalog as well.

I'm going to try seeing if I can catch what each of the main catalogs throw as the exception now and then re-throw it according to the current interface to keep `NoSuchNamespaceExeption`. An exception comes up in all cases, it's just not that.

Reply:
> What if user drop same namespace cascade concurrently? Maybe we should keep NoSuchNamespaceException

I managed to get `NoSuchNamespaceException` back. Will push shortly.

The general case of concurrent removals / changes won't be made worse by this PR. But any further hardening the catalog around concurrent access via synchronize or some other means should be done via another ticket in my opinion to separate concerns.